src/database-sqlite3.cpp

```diff
@@ -56,6 +65,50 @@ SQLite format specification:
 			 sqlite3_errmsg(m_database)); \
 	}
 
+static int sqlite3BusyHandler(void *data, int count)
+{
+	s64 &first_time = reinterpret_cast<s64 *>(data)[0];
+	s64 &prev_time = reinterpret_cast<s64 *>(data)[1];
+	s64 cur_time = getTimeMs();
+
+	if (count == 0)
+		first_time = cur_time;
+	else {
+		while (cur_time < prev_time)
+			cur_time += s64(1)<<32;
+	}
+
+	if (cur_time - first_time < BUSY_INFO_TRESHOLD_1) {
+		; // do nothing
+	} else if (cur_time - first_time >= BUSY_INFO_TRESHOLD_1 &&
+			prev_time - first_time < BUSY_INFO_TRESHOLD_1) {
+		infostream << "NOTE: sqlite3 database has been locked for "
+			<< cur_time - first_time << " ms." << std::endl;
+	} else if (cur_time - first_time >= BUSY_INFO_TRESHOLD_2 &&
+			prev_time - first_time < BUSY_INFO_TRESHOLD_2) {
+		infostream << "NOTE: sqlite3 database has been locked for "
+			<< cur_time - first_time << " ms." << std::endl;
+	} else if (cur_time - first_time >= BUSY_ERROR_TRESHOLD &&
+			prev_time - first_time < BUSY_ERROR_TRESHOLD) {
+		errorstream << "WARNING: sqlite3 database has been locked for "
+			<< cur_time - first_time << " ms; this causes lag." << std::endl;
+	} else if (cur_time - first_time >= BUSY_FATAL_TRESHOLD &&
+			prev_time - first_time < BUSY_FATAL_TRESHOLD) {
+		errorstream << "ERROR: sqlite3 database has been locked for "
+			<< cur_time - first_time << " ms - giving up!" << std::endl;
+	} else if ((cur_time - first_time) / BUSY_ERROR_INTERVAL !=
+			(prev_time - first_time) / BUSY_ERROR_INTERVAL) {
+		// Safety net: keep reporting every BUSY_ERROR_INTERVAL
```

Comment:
It seems weird to have all this code dedicated to cleanly handle an impossible situation.  Why not just `FATAL_ERROR("Passed maximum SQLite3 database wait time.")`?